### ISSUE TITLE
[avmfritz] Add call monitor support

### ISSUE THREAD

**Issue by kaikreuzer** [openhab/openhab-addons#8167]
The documentation at https://www.openhab.org/addons/bindings/avmfritz/ states:
" It is planned to support some channels directly at the bridge like call monitoring and others - but these things are under current development by now."

For me, the call monitor is a missing feature for openHAB 3, so I'd be very interested to seeing this added.
@cweitkamp You are listed as CODEOWNER for this binding. Are you actively using it and would you be the right person to discuss whether this fits in here and if so, how?

**Comment by cweitkamp** [comment]
Yes, I am using the Call monitor of the v1 fritzboxtr064 binding - at least a little bit - as well as the avmfritz binding and I am still maintaining it.

AFAIK the Call monitor and TAM features are based on the TR064 protocol whereas the avmfritz binding uses a different interface ([AHA-HTTP-Interface](https://avm.de/fileadmin/user_upload/Global/Service/Schnittstellen/AHA-HTTP-Interface.pdf)) to communicate with the FRITZ!Box. I am not sure if the right place to add these features here even if it sound reasonable. @J-N-K would it be possible to reuse the classes / interfaces provided by the tr064 introduced in #6678?

**Comment by kaikreuzer** [comment]
The new TR064 binding is really about supporting that protocol and not a dedicated "FritzBox" binding anymore.
And no, afaik, the call monitor is not part of TR064 (and I do not see it listed [here](https://github.com/openhab/openhab-addons/pull/6678/files#diff-bd6d8867117e6e440b76d9afb6b5d204R62-R95)).
In openHAB 1, there is the [dedicated binding](https://www.openhab.org/addons/bindings/fritzbox1/#fritz-box-binding), which uses telnet communication for call monitoring - that's the one I am using and which I need. As I found it mentioned in the avmfritz readme, I thought that's indeed the best place to move it to then (I would not want to add yet another Fritzbox binding just for the call monitor). What I am now confused about is the statement "support for enabling telnet on the FRITZ!Box has been dropped due to security reasons in firmware version 06.25 and later" on https://www.openhab.org/addons/bindings/fritzbox1/#fritz-box-binding. Does this mean that this feature is not supported anymore by AVM?

**Comment by kaikreuzer** [comment]
> Does this mean that this feature is not supported anymore by AVM?

Browsing the web, I cannot find any trace of that - I'd hence assume that the statement is not correct and we should be fine going with the telnet communication.
The only "nasty" thing about adding it to the avmfritz binding is that the code introduces a dependency on `org.apache.commons.net.telnet`. Not sure, if we find any suitable alternative, but I would say that it would be acceptable.

**Comment by kaikreuzer** [comment]
I just saw that indeed the 1.x TR064 binding [has call monitor support](https://www.openhab.org/addons/bindings/fritzboxtr0641/#features) as well. But this is not done through the TR064 protocol, but through telnet as well. The good thing is that [its implementation](https://github.com/openhab/openhab1-addons/blob/master/bundles/binding/org.openhab.binding.fritzboxtr064/src/main/java/org/openhab/binding/fritzboxtr064/internal/CallMonitor.java) is done without a dependency to Apache Commons Telnet.

Still, I would claim that the feature would be a better fit here in "avmfritz" (as it is not called "avmaha" anymore and thus suggests to have further features besides aha implemented as well).

@J-N-K and @cweitkamp Would you agree? If so, I volunteer to do a PR for porting that feature.

**Comment by J-N-K** [comment]
Yes. I originally thought about implementing it in TR064, but decided against it because it is a completely different protocol.

I think AVM Fritz would be a good choice. We should check before that „modern“ FritzOs versions still support it. If I‘m not mistaken 6.25 is quite old.

**Comment by cweitkamp** [comment]
Okay, I was not aware of the dedicated binding for it. I am using old fritzboxtr064 only without remembering to enable telnet. 

It looks like the Callmonitor is kind of hidden feature of AVM. At least there is no real documentation available. What I can tell you is that it is still working on my F!B 7590 with F!OS 7.20 (latest OS version). I have the slight feeling they decoupled telnet and Callmonitor in an earlier OS version. Callmonitor is available at TCP port 1012 and can be activated by dialing `#96*4*` (`#96*5*` for deactivation). The telnet server is still present in background but deactivated by default and not needed for Callmonitor anymore (e.g. dial `#96*6*` to enable it, `#96*7*` to disable it). 

I am fine with moving it to avmfritz binding. No objections from my side.

**Comment by kaikreuzer** [comment]
> We should check before that „modern“ FritzOs versions still support it.

I can also confirm that I can still use this feature on my FB7590 with latest firmware.
@cweitkamp is right - the feature is actually not related to telnet, but it is just simple TCP on port 1012.

I'll come up with a PR - thank you both!

**Comment by cweitkamp** [comment]
Great. I assigned thus issue to yourself. Let me know once you have a working version. I am happy to help testing it.

**Comment by kaikreuzer** [comment]
Hmph... I started looking into it and directly noticed an issue: The call monitor feature in the 1.x binding is directly linked to TR064 since it uses the phonebook for looking up names for numbers.
If we add the call monitor to the avmfritz binding, we'd clearly only be able to provide phone numbers, but no caller names.
Adding it to the TR064 binding would mean that it becomes pretty AVM specific, which we tried to avoid.
From a user perspective, I'd actually think it would be best to combine all features in the avmfritz binding, so that there's only a single FB-thing that needs to be configured with credentials.
So maybe we should take a step back and rediscuss our general direction here. Without digging deeper into it, I wonder whether moving tr064 to be an io.transport.tr064 bundle and using it directly in the avmfritz binding could be an option? As many features are done as sub-things, we could still have a clear separation of the code in different handler classes.

@cweitkamp & @J-N-K What's your take on this on how to best proceed?

**Comment by DerOetzi** [comment]
Hello there,

I although use old fritzboxtr064 binding for call monitoring of my 7590 (actual firmware 7.20) I can offer to do some investigation on this as well.

Regards

**Comment by J-N-K** [comment]
We could also add an action to lookup the phone number from the tr064 binding in a phonebook provided by the avmfritz binding. If the phonebook is a `Map<Number, Name>`, this could also be generated by other sources than the avmfritz binding (an SQL database, a HTTP request).

**Comment by DerOetzi** [comment]
I have done some investigation and found a solution to download the phonebook as xml by a simple http-post-request. All we need is a valid Session ID which we already have in avmfritz binding if I see it correct.

Send the post request to http://fritz.box/cgi-bin/firmwarecfg with following Post-Data:

```
sid: __SESSION_ID__,
PhonebookId: 0
PhonebookExportName: Telefonbuch
PhonebookExport: 
```

**Comment by kaikreuzer** [comment]
> If the phonebook is a Map<Number, Name>, this could also be generated by other sources than the avmfritz binding (an SQL database, a HTTP request).

@J-N-K I do not quite understand this comment. My understanding so far was that reading the phonebook was a feature of TR064, so what would maybe work is to create an action that does a name lookup for a given number. Nonetheless, from a user perspective I would think it would be much nicer to have that within a single avmfritz binding instead of having to combine features of several bindings through a rule.

**Comment by J-N-K** [comment]
@kaikreuzer True. Implementing the phonebook lookup looks easy. Most functionality is already present from the caller lists. Of course this will only be available when it is a AVM Fritzbox. Since it already took quite some time to refactor and get it working again from fritzboxtr064 to tr064 I would kindly ask not to rename or move the bundle again.

**Comment by kaikreuzer** [comment]
> download the phonebook as xml by a simple http-post-request

@DerOetzi Yes, this seems to be possible, but it seems to me that this is rather an inofficial way to trigger an export of the phonebook. The TR064 way is instead the [officially documented](https://avm.de/fileadmin/user_upload/Global/Service/Schnittstellen/x_contactSCPD.pdf) (and thus recommended) API for it.
If we want to have the caller name lookup in avmfritz without requiring the user to also use the tr064 binding, it would be an option to consider, though.

@J-N-K So your suggestion would be to implement the call monitor in the avmfritz binding and use an action of the tr064 binding to do the caller name resolution, correct?

I still feel that this is a step back for the user that is coming from the 1.x version... And looking at the current avmfritz binding, it really is all about AHA and there are no channels at all for the FB itself - everything that is around DSL/Wifi/Calls/etc. is in the TR064 binding, so from a functional pov, I would claim that it would be the better fit and we could implement it there. Still, users will likely look at the "avmfritz" binding first for all such features... I'm really torn apart on how to best proceed.

> Since it already took quite some time to refactor and get it working again from fritzboxtr064 to tr064 I would kindly ask not to rename or move the bundle again.

I very much appreciate this and clearly understand your request. Nonetheless, it is now the only chance to decide on the way forward and if our discussion leads us to a solution that we all think is best, we should not be held back from implementing it. We all didn't have the call monitor in scope before and I only start understanding our fragmented FB support now after diving deeper into it...
But as you and @cweitkamp are the maintainers of the two respective bindings, I will listen to what you think is best, also taking technical considerations into account. I'd hence also hope for @cweitkamp to share his perspective on that.

**Comment by cweitkamp** [comment]
From my POV the separation of TR064 and AVM specific features makes senses and is the right way to go. Currently I am not aware of other services depending on TR064 but as it is a common standard it would be a good choice to introduce a transport bundle for it which could be used by different bindings.

I agree on the opinion to merge AVM specific features into one binding regardless of their origin. It is a clear approach and simplifies understanding for users.

Regarding the feature to resolve numbers to caller names I like the idea of @J-N-K to expose it for other sources. But I think it is more a mapping for numbers to names and thus imo it better fits into a Profile (not only for phonebooks available via TR064). Maybe some day the Android or iOS App will reflect the calling number to an openHAB Item and then you can resolve the related name via it or use Google Contacts or similar APIs.

**Comment by kaikreuzer** [comment]
> From my POV the separation of TR064 and AVM specific features makes senses

Note though that a lot of AVM specific features are done through TR064, like e.g. the phone book. There hence isn't really any clean separation possible.

> as it is a common standard it would be a good choice to introduce a transport bundle for it which could be used by different bindings.

I am not yet deep enough in TR064 to have an opinion on this. I understand that TR064 is a kind of extension of UPnP and as such it is pretty generic. Not sure how much of the functionality could be extracted into a transport bundle, which is not just plain UPnP + SOAP handling. I guess only @J-N-K can tell.

> I agree on the opinion to merge AVM specific features into one binding regardless of their origin

So what does this mean for the phonebook for you? As mentioned above, it is a custom AVM feature, but realized through TR064 (hence it will be part of the TR064 binding according to the current plan).

> more a mapping for numbers to names and thus imo it better fits into a Profile

A profile is indeed a good match. It would be still quite easy to use without having to write any rule. 👍 

> Maybe some day the Android or iOS App will reflect the calling number to an openHAB Item and then you can resolve the related name

This won't work since profiles are related to links, so they will only convert what is coming from a channel. But we could still also offer an action along with the profile for such cases.

**Comment by cweitkamp** [comment]
> that TR064 is a kind of extension of UPnP and as such it is pretty generic

Yes, that is my understanding too. But I am not experienced enough to tell you which specifications belong to generic stuff and which are AVM specific.

> what does this mean for the phonebook for you?

If it is possible to have a dedicated transport bundle I would prefer to see the phonebook in avmfritz binding. If that is not possible and we go for a tr064 binding covering AVM specific features too we should place it over there and setup the ways to solve the lookup.

To be clear. My favorite solution is a transport bundle to be used by a universal tr064 binding and the existing avmfritz binding.

**Comment by J-N-K** [comment]
The tricky part in the binding is not the protocol, that's a very simple SOAP request. The difficult part is to determine what channels are available and how to read or write them. This information is available as SCPD (an XML file which contains the information about what is called actions - like `EnableWifi` - and the parameters needed for that and the datatype for these parameters).

The OH1 TR064 had this information hardcoded, which would lead to all channels available but not all of them working. Parsing the SCPD file allows to create the channels dynamically, depending on their availability and configuration. It also allows finding the sub-devices (i.e. the LAN, WAN, WLAN interfaces) and their parameters. This is clearly thing-related. I have no idea how this could be moved to a transport bundle, without duplication the channel creating in each bundle using that transport bundle.

Adding the `PHONEBOOK_LOOKUP` profile is doable. We could chose to use either a specified phonebook (you can define more than one) or all phonebooks available.

**Comment by kaikreuzer** [comment]
Ok then. I have started working on a PR (https://github.com/openhab/openhab-addons/pull/8220), which will add the "number only" call monitor to the avmfritz binding.

**Comment by DerOetzi** [comment]
Another point I want to add to discussion. First of all I'm not sure about this belongs to call monitor or is a classic tr064 use case. I use at the moment for rule triggering  `{fritzboxtr064="missedCallsInDays:2"}` Maybe the binding can identify missing calls by the order of events RING->CALL->DISCONNECT vs. RING->DISCONNECT and trigger an event-channel

**Comment by J-N-K** [comment]
Last caller is a tr064 feature.

**Comment by DerOetzi** [comment]
> 
> 
> Last caller is a tr064 feature.

Agree for last caller. But I still think about the pure functionality of identify a call was picked up or missed. I think this can be handled although without having user to install two different bindings.
